packages/doenetml-worker/lib-doenetml-core/src/dast/extended_vec.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    ops::{Index, IndexMut},
+    slice::Iter,
+};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ExtendedVector<T> {
+    inner: Vec<T>,
+    index_offset: isize,
+}
+
+impl<T> ExtendedVector<T> {
+    pub fn new(index_offset: isize) -> ExtendedVector<T> {
+        ExtendedVector {
+            inner: Vec::new(),
+            index_offset,
+        }
+    }
+
+    pub fn from_vector(values: Vec<T>, index_offset: isize) -> ExtendedVector<T> {
+        ExtendedVector {
+            inner: values,
+            index_offset,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    pub fn insert(&mut self, index: isize, element: T) {
+        if index + self.index_offset < 0 {
+            panic!("Extended vector index cannot be less than index offset");
+        }
+        self.inner
+            .insert((index + self.index_offset).try_into().unwrap(), element);
```

Comment:
I guess there is no `into()`? You can also do `as usize`. I am not sure which is better.

Reply:
code removed